File: components/users/src/com/github/joerodriguez/sbng2ex/UserService.java
package com.github.joerodriguez.sbng2ex;

import com.github.joerodriguez.sbng2ex.service.ErrorType;
import com.github.joerodriguez.sbng2ex.service.ServiceResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.regex.Pattern;

@Component
public class UserService {

    public static final Pattern EMAIL_PATTERN = Pattern.compile("^(.+)@(.+)$");

    private final UsersRepository usersRepository;

    @Autowired
    public UserService(UsersRepository usersRepository) {
        this.usersRepository = usersRepository;
    }

    public ServiceResponse<User> create(String email, String password) {
        return ServiceResponse.create(response -> {

            if (!EMAIL_PATTERN.matcher(email).matches()) {
                response.addError(ErrorType.INVALID_EMAIL.forField("email"));
            }

            if (usersRepository.findByEmail(email).isPresent()) {
                response.addError(ErrorType.EMAIL_TAKEN.forField("email"));
            }

            response.apply(() -> usersRepository.create(email, password));
        });
    }
}

File: components/users/src/com/github/joerodriguez/sbng2ex/UserService.kt
package com.github.joerodriguez.sbng2ex

import com.github.joerodriguez.sbng2ex.service.ErrorType
import com.github.joerodriguez.sbng2ex.service.ServiceResponse
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import java.util.regex.Pattern

@Component
class UserService

    @Autowired
    constructor(private val usersRepository: UsersRepository)

{

    private val emailPattern = Pattern.compile("^(.+)@(.+)$")

    fun create(email: String, password: String) = ServiceResponse.create<User> {

        if (!emailPattern.matcher(email).matches()) {
            it.error(ErrorType.INVALID_EMAIL.forField("email"))
        }

        if (usersRepository.findByEmail(email) != null) {
            it.error(ErrorType.EMAIL_TAKEN.forField("email"))
        }

        it.apply { usersRepository.create(email, password) }
    }

}
